### DISCUSSION TITLE
Shift in energy test set and pressure

### DISCUSSION THREAD

**Discussion by cesaremalosso** [deepmodeling/deepmd-kit#2935] [Q&A]
### Summary

Shift in energy test set and pressure when I run it on a dataset with different number of atoms with respect to the training one.

### DeePMD-kit Version

v2.1.1.dev80+g8c58455

### TensorFlow Version

2.3.0

### Details

Hi, I trained a NN over a data set made of snapshots a 432 atoms system. When I test this NN over a different dataset made of 576 atoms ( of course not included in the train one) I find good predictions for forces, so the dynamics of the system should be fine, but I find a rigid shift upon the total energy and also on the diagonal elements of the viral.

energy
![IMAGE 2022-12-14 18:28:48](https://user-images.githubusercontent.com/72449246/207665604-b3dda870-df32-4045-bc73-9140f54873fc.jpg)

yy diagonal element of the virial
<img width="1104" alt="Screenshot 2022-12-14 at 18 29 49" src="https://user-images.githubusercontent.com/72449246/207666118-d63710ea-069b-486a-819e-35458f7c6f5e.png">

On the training set these shifts are not present. I thought that specifying the atomic energies of each species would have solve the problem, but It did not. Is there any kind of solution to this problem?

Thanks

**Comment by njzjz** [comment]
Do your two datasets have the same chemical composition?

**Comment by cesaremalosso** [comment]
Yeah, it is NH3

**Comment by wanghan-iapcm** [comment]
The shift in virial is not expected. Could you please check, by your labeling method, if the virial is equal to the derivative of the energy?

**Comment by cesaremalosso** [comment]
Do you mean that you want me to check if the first-principles calculations (made with quantum espresso) are correct? I've already checked the convergence of the viral tensor and it's converged with respect to the plane waves cut off that I used. I've also noticed that training on the virial seems to reduce this shift but I was wondering why is it a rigid shift and not just a random error of the NN. It seems like there is some kind of bias somewhere.

**Comment by wanghan-iapcm** [comment]
You may check if the following equation holds
```math
P_{\alpha\beta} = - \frac{1}{\det(h)} \sum_{\gamma=1}^3 h_{\beta\gamma} \frac{\partial E}{\partial h_{\alpha\gamma}}
```
where $h_{\alpha\gamma}$ is the $\gamma$-th component of the $\alpha$-th cell vector.

**Comment by njzjz** [comment]
Can you provide an example for reproducing?

**Comment by cesaremalosso** [comment]
Here's the [link](https://www.dropbox.com/s/7bnlz6pmbmjofod/git_issue.tar?dl=0) 
In the link you find a directory that contains three different directories:
- `model_training`: contains the model and the training dataset (generated with DPGEN) and a directory called `test_dataset` that contains data taken from an _ab-initio_ MD simulation (runned with quantumESPRESSO);
- `scf_qe`: contains an example of a first-principle calculation (quantumESPRESSO);
- `dp_test` contains the test over the `test_dataset`;

**Comment by njzjz** [comment]
I plot `dp_test/test.e.out` and obtain a different figure:

![Screenshot from 2022-12-21 17-25-31](https://user-images.githubusercontent.com/9496702/209015067-23794a57-d853-4376-825c-a2d9c7a683ad.png)

In this figure, the bias is about 5 eV for 576 atoms instead of 15 eV in your figure.

Note that your training data has a similar bias.

**Comment by cesaremalosso** [comment]
Yes, you are right, the model that I linked you is trained over 2 million steps while the one of the plots over 1 million steps. I [LINK](https://www.dropbox.com/s/whnda9rmw5vajkx/git_issue1.tar?dl=0) you the correct model.
aGAIN:
- `model_training`: contains the model and the training dataset (generated with DPGEN) and a directory called `test_dataset `that contains data taken from an _ab-initio_ MD simulation (runned with quantumESPRESSO);
- `scf_qe`: contains an example of a first-principle calculation (quantumESPRESSO);
- `dp_test` contains the test over the `test_dataset` and over some of the training data .

You can see that theres is a shift both in energy and virial (diagonal) in the test_data while there's not in the train_data. Is this only a matter of the quality of the training, in the sense that increasing the number of training steps would solve the problem? Why is the error such a rigid shift in the diagonal terms of the virial? Why are the off-diagonal element of the virial instead correct?

**Comment by IricsDo** [comment]
Hi everyone, I get the same problem. Seem like the energy shift (has the gap) between predict and dft (true lable). Have you found the solution yet?
Edit: I get the same problem when I draw with virial value!
![image](https://github.com/user-attachments/assets/3db2041c-9594-48d2-b0dd-6651524b6203)
